src/ugc.rs

```diff
@@ -0,0 +1,535 @@
+            sys::SteamAPI_ISteamUGC_SetReturnChildren(self.ugc, self.handle, include)
+        };
+        debug_assert!(ok);
+        self
+    }
+
+    /// Include metadata in results
+    pub fn include_metadata(self, include: bool) -> Self {
+        let ok = unsafe {
+            sys::SteamAPI_ISteamUGC_SetReturnMetadata(self.ugc, self.handle, include)
+        };
+        debug_assert!(ok);
+        self
+    }
+
+    /// Include additional previews in results
+    pub fn include_additional_previews(self, include: bool) -> Self {
+        let ok = unsafe {
+            sys::SteamAPI_ISteamUGC_SetReturnAdditionalPreviews(self.ugc, self.handle, include)
+        };
+        debug_assert!(ok);
+        self
+    }
+
+    /// Runs the query
+    pub fn fetch<F>(self, mut cb: F)
+        where F: FnMut(Result<QueryResults,SteamError>) + 'static + Send
+    {
+        let ugc = self.ugc;
+        let inner = Arc::clone(&self.inner);
+        let handle = self.handle;
+        mem::forget(self); // Don't run destructor since we need the handle still
+
+        unsafe {
+            let api_call = sys::SteamAPI_ISteamUGC_SendQueryUGCRequest(ugc, handle);
+            register_call_result::<sys::SteamUGCQueryCompleted_t, _, _>(
+                &inner, api_call, CALLBACK_BASE_ID + 1,
+                move |v, io_error| {
+                    if io_error {
+                        cb(Err(SteamError::IOFailure));
+                        return;
+                    } else if v.m_eResult != sys::EResult::EResultOK {
+                        cb(Err(v.m_eResult.into()));
+                        return;
+                    }
+
+                    let result = QueryResults {
+                        ugc: sys::steam_rust_get_ugc(),
+                        handle,
+                        num_results_returned: v.m_unNumResultsReturned,
+                        num_results_total: v.m_unTotalMatchingResults,
+                        was_cached: v.m_bCachedData,
+                    };
+                    cb(Ok(result));
+            });
+        }
+    }
+
+    /// Runs the query, only fetching the total number of results.
+    pub fn fetch_total<F>(self, mut cb: F)
+        where F: FnMut(Result<u32, SteamError>) + 'static + Send
+    {
+        unsafe {
+            let ok = sys::SteamAPI_ISteamUGC_SetReturnTotalOnly(self.ugc, self.handle, true);
+            debug_assert!(ok);
+        }
+
+        self.fetch(move |res| cb(res.map(|qr| qr.total_results())))
+    }
+
+    /// Runs the query, only fetchind the IDs.
+    pub fn fetch_ids<F>(self, mut cb: F)
+        where F: FnMut(Result<Vec<u64>, SteamError>) + 'static + Send
+    {
+        unsafe {
+            let ok = sys::SteamAPI_ISteamUGC_SetReturnOnlyIDs(self.ugc, self.handle, true);
+            debug_assert!(ok);
+        }
+
+        self.fetch(move |res| cb(res.map(|qr| qr.iter().map(|v| v.published_file_id).collect::<Vec<_>>())))
+    }
+}
+
+/// Query results
+pub struct QueryResults {
```

Comment:
This should really contain either a `Arc<Inner<Manager>>` or a lifetime to prevent the pointer from being invalid if the steamworks client is dropped.

Looking at the call to `SteamAPI_ISteamUGC_GetQueryUGCResult` below which requires it to be only called during the callback above a lifetime would be better or just passing it as a reference to the callback function. Just need to prevent it from being used outside of that callback.

Reply:
`Arc<Inner<Manager>>` isn't `Send` iirc. I think I did have that as a field once but it didn't compile. A lifetime parameter could work though.